File: packages/agenti-mcp-server/packages/infra/mcp-notify/pkg/x402/middleware.go
// Package x402 provides payment gating middleware for MCP servers
// using the x402 payment protocol.
package x402

import (
	"encoding/json"
	"fmt"
	"os"
	"sync"
	"time"
)

// Config holds x402 payment configuration
type Config struct {
	Price       string `json:"price"`
	Token       string `json:"token"`
	Chain       string `json:"chain"`
	Recipient   string `json:"recipient"`
	Description string `json:"description,omitempty"`
	FreeTier    struct {
		Calls  int `json:"calls"`
		Period int `json:"period"` // seconds
	} `json:"freeTier"`
}

// PaymentProof represents a verified payment
type PaymentProof struct {
	Signature string  `json:"signature"`
	Timestamp int64   `json:"timestamp"`
	Amount    string  `json:"amount"`
	Token     string  `json:"token"`
	Payer     string  `json:"payer"`
}

// PaymentRequired represents a 402 response
type PaymentRequired struct {
	Error struct {
		Code    int    `json:"code"`
		Message string `json:"message"`
		X402    struct {
			Version     string   `json:"version"`
			Price       string   `json:"price"`
			Token       string   `json:"token"`
			Chain       string   `json:"chain"`
			Recipient   string   `json:"recipient"`
			Description string   `json:"description,omitempty"`
			Accepts     []string `json:"accepts"`
		} `json:"x402"`
	} `json:"error"`
}

// Middleware handles x402 payment verification
type Middleware struct {
	config     Config
	usageCache map[string]*usage
	mu         sync.RWMutex
}

type usage struct {
	count       int
	periodStart time.Time
}

// DefaultConfig returns default x402 configuration
func DefaultConfig() Config {
	recipient := os.Getenv("X402_RECIPIENT")
	if recipient == "" {
		recipient = "0x1234567890123456789012345678901234567890"
	}
	
	return Config{
		Price:     "0.001",
		Token:     "USDC",
		Chain:     "base",
		Recipient: recipient,
		FreeTier: struct {
			Calls  int `json:"calls"`
			Period int `json:"period"`
		}{
			Calls:  10,
			Period: 3600,
		},
	}
}

// New creates a new x402 middleware
func New(config Config) *Middleware {
	return &Middleware{
		config:     config,
		usageCache: make(map[string]*usage),
	}
}

// CheckFreeTier checks if client is within free tier limits
func (m *Middleware) CheckFreeTier(clientID string) bool {
	m.mu.Lock()
	defer m.mu.Unlock()
	
	now := time.Now()
	u, exists := m.usageCache[clientID]
	
	if !exists || now.Sub(u.periodStart).Seconds() > float64(m.config.FreeTier.Period) {
		m.usageCache[clientID] = &usage{
			count:       1,
			periodStart: now,
		}
		return true
	}
	
	if u.count < m.config.FreeTier.Calls {
		u.count++
		return true
	}
	
	return false
}

// VerifyPayment verifies a payment proof
func (m *Middleware) VerifyPayment(proof *PaymentProof) bool {
	if proof == nil {
		return false
	}
	
	// Check timestamp (within 5 minutes)
	now := time.Now().Unix()
	if abs(now-proof.Timestamp) > 300 {
		return false
	}
	
	// Check token
	if proof.Token != m.config.Token {
		return false
	}
	
	// In production, verify cryptographic signature
	return true
}

// CreatePaymentRequired creates a 402 response
func (m *Middleware) CreatePaymentRequired() PaymentRequired {
	var resp PaymentRequired
	resp.Error.Code = 402
	resp.Error.Message = "Payment Required"
	resp.Error.X402.Version = "1.0"
	resp.Error.X402.Price = m.config.Price
	resp.Error.X402.Token = m.config.Token
	resp.Error.X402.Chain = m.config.Chain
	resp.Error.X402.Recipient = m.config.Recipient
	resp.Error.X402.Description = m.config.Description
	resp.Error.X402.Accepts = []string{"x402-payment"}
	return resp
}

// PricingInfo returns a pricing description string
func (m *Middleware) PricingInfo() string {
	return fmt.Sprintf("💰 %s %s per call (%s)", m.config.Price, m.config.Token, m.config.Chain)
}

// ToJSON converts config to JSON
func (c Config) ToJSON() ([]byte, error) {
	return json.Marshal(c)
}

func abs(n int64) int64 {
	if n < 0 {
		return -n
	}
	return n
}

File: packages/agenti-mcp-server/packages/generators/ucai/src/abi_to_mcp/x402/middleware.py
"""
x402 Payment Middleware for UCAI MCP Servers

Provides payment gating for MCP tools using the x402 payment protocol.
"""

import os
import json
import time
from dataclasses import dataclass, field
from typing import Optional, Callable, Any, Dict
from functools import wraps


@dataclass
class X402Config:
    """Configuration for x402 payment integration."""
    
    price: str = "0.001"
    token: str = "USDC"
    chain: str = "base"
    recipient: str = field(default_factory=lambda: os.environ.get(
        "X402_RECIPIENT", 
        "0x1234567890123456789012345678901234567890"
    ))
    free_tier_calls: int = 10
    free_tier_period: int = 3600
    description: Optional[str] = None
    
    def to_dict(self) -> Dict[str, Any]:
        return {
            "price": self.price,
            "token": self.token,
            "chain": self.chain,
            "recipient": self.recipient,
            "description": self.description,
        }


class X402Middleware:
    """Middleware for handling x402 payments in MCP servers."""
    
    def __init__(self, config: Optional[X402Config] = None):
        self.config = config or X402Config()
        self._usage_cache: Dict[str, Dict[str, Any]] = {}
    
    def _check_free_tier(self, client_id: str) -> bool:
        now = time.time()
        
        if client_id not in self._usage_cache:
            self._usage_cache[client_id] = {
                'count': 0,
                'period_start': now,
            }
        
        usage = self._usage_cache[client_id]
        
        if now - usage['period_start'] > self.config.free_tier_period:
            usage['count'] = 0
            usage['period_start'] = now
        
        if usage['count'] < self.config.free_tier_calls:
            usage['count'] += 1
            return True
        
        return False
    
    def _verify_payment_header(self, payment_header: Optional[str]) -> bool:
        if not payment_header:
            return False
        
        try:
            payment_data = json.loads(payment_header)
            required_fields = ['signature', 'timestamp', 'amount', 'token']
            if not all(f in payment_data for f in required_fields):
                return False
            if abs(time.time() - payment_data['timestamp']) > 300:
                return False
            if payment_data['token'] != self.config.token:
                return False
            if float(payment_data['amount']) < float(self.config.price):
                return False
            return True
        except (json.JSONDecodeError, KeyError, ValueError):
            return False
    
    def create_payment_required_response(self) -> Dict[str, Any]:
        return {
            "error": {
                "code": 402,
                "message": "Payment Required",
                "x402": {
                    "version": "1.0",
                    "price": self.config.price,
                    "token": self.config.token,
                    "chain": self.config.chain,
                    "recipient": self.config.recipient,
                    "description": self.config.description,
                    "accepts": ["x402-payment"],
                }
            }
        }


def create_payment_required_response(config: X402Config) -> Dict[str, Any]:
    middleware = X402Middleware(config)
    return middleware.create_payment_required_response()


def verify_payment(payment_header: Optional[str], config: X402Config) -> bool:
    middleware = X402Middleware(config)
    return middleware._verify_payment_header(payment_header)


def with_x402(
    config: Optional[X402Config] = None,
    price: Optional[str] = None,
    token: Optional[str] = None,
    description: Optional[str] = None,
):
    """Decorator to add x402 payment gating to an MCP tool."""
    if config is None:
        config = X402Config(
            price=price or "0.001",
            token=token or "USDC",
            description=description,
        )
    
    middleware = X402Middleware(config)
    
    def decorator(func: Callable) -> Callable:
        @wraps(func)
        async def wrapper(*args, **kwargs):
            payment_header = kwargs.pop('_x402_payment', None)
            request = kwargs.get('request', None)
            client_id = 'anonymous'
            if request and hasattr(request, 'client_id'):
                client_id = request.client_id
            
            if middleware._check_free_tier(client_id):
                return await func(*args, **kwargs)
            
            if middleware._verify_payment_header(payment_header):
                return await func(*args, **kwargs)
            
            return middleware.create_payment_required_response()
        
        wrapper._x402_config = config
        wrapper._x402_pricing = config.to_dict()
        return wrapper
    
    return decorator


def pricing_info(config: X402Config) -> str:
    return f"💰 {config.price} {config.token} per call ({config.chain})"
